applications/ShapeOptimizationApplication/custom_utilities/damping/damping_utilities.h

```diff
@@ -0,0 +1,396 @@
+#include "../../kratos/utilities/binbased_fast_point_locator.h"
+#include "../../kratos/utilities/normal_calculation_utils.h"
+#include "shape_optimization_application.h"
+#include "damping_function.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+    typedef boost::python::extract<ModelPart&> extractModelPart;
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+
+*/
+
+class DampingUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    // Type definitions for better reading later
+    typedef array_1d<double,3> array_3d;
+    typedef Node < 3 > NodeType;
+    typedef Node < 3 > ::Pointer NodeTypePointer;
+    typedef std::vector<NodeType::Pointer> NodeVector;
+    typedef std::vector<NodeType::Pointer>::iterator NodeIterator;
+    typedef std::vector<double> DoubleVector;
+    typedef std::vector<double>::iterator DoubleVectorIterator;
+    typedef ModelPart::ConditionsContainerType ConditionsArrayType;
+
+    // Type definitions for tree-search
+    typedef Bucket< 3, NodeType, NodeVector, NodeTypePointer, NodeIterator, DoubleVectorIterator > BucketType;
+    typedef Tree< KDTreePartition<BucketType> > KDTree;    
+
+    /// Pointer definition of DampingUtilities
+    KRATOS_CLASS_POINTER_DEFINITION(DampingUtilities);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    DampingUtilities( ModelPart& modelPartToDamp, boost::python::dict subModelPartsForDamping, Parameters& optimizationSettings )
+        : mrModelPartToDamp( modelPartToDamp ),
+          mrDampingRegions( subModelPartsForDamping ),
+          mDampingSettings( optimizationSettings["design_variables"]["damping"]["damping_regions"] )
+    {
+        boost::timer timer;        
+        std::cout << "\n> Creating search tree to perform damping..." << std::endl;
+        CreateListOfNodesOfModelPart();
+        CreateSearchTreeWithAllNodesOfModelPart();
+        std::cout << "> Search tree created in: " << timer.elapsed() << " s" << std::endl;
+
+        InitalizeDampingFactorsToHaveNoInfluence(); 
+        SetDampingFactorsForAllDampingRegions();         
+    }
+
+    /// Destructor.
+    virtual ~DampingUtilities()
+    {
+    }
+
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    // ==============================================================================
+    void CreateListOfNodesOfModelPart()
+    {
+        for (ModelPart::NodesContainerType::iterator node_it = mrModelPartToDamp.NodesBegin(); node_it != mrModelPartToDamp.NodesEnd(); ++node_it)
+        {
+            NodeTypePointer pnode = *(node_it.base());
+            mListOfNodesOfModelPart.push_back(pnode);
```

Comment:
resizing the mListOfNodesOfModelPart list before the loop would be more efficient than using push_back